README.rst
===============================================
rebus
===============================================
Generate base64-encoded strings consisting of alphanumeric symbols only.

Why did I make this app?
===============================================
Recently I have been working on implementing of a Google Authenticator app for
Plone (which I did for my beloved company - Goldmund, Wyldebeast & Wunderliebe).

For generating of a bar code image, I needed a base32 encoded string. While Android
devices could perfectly scan all bar-code image that I would generate, Apple devices
would raise errors on bar code images which were generated using strings that contain
one or more "=" characters.

The solution found was to add a number of \\n at the end of the string to be encoded.

If you happen to experience similar problems, you know what to do.

Prerequisites
===============================================
- Python 2.6.8+, 2.7.+, 3.3.+

Installation
===============================================
Install latest stable version from PyPI:

    $ pip install rebus

...or latest stable version from GitHub:

    $ pip install -e git+https://github.com/barseghyanartur/rebus@stable#egg=rebus

...or latest stable version from BitBucket:

    $ pip install -e hg+https://bitbucket.org/barseghyanartur/rebus@stable#egg=rebus


Usage and examples
===============================================
Using `rebus` is damn easy. Whenever you would want to use ``base64.b32encode``
or ``base64.b64encode``, replace `base64` with `rebus`.

Encoding
-----------------------------------------------
Required imports

>>> import rebus

b32encode string

>>> rebus.b32encode('abcdefg')
'MFRGGZDFMZTQUCQK'

b64encode string

>>> rebus.b64encode('abcdefg')
'YWJjZGVmZwoK'

urlsafe_b64encode

>>> rebus.urlsafe_b64encode('abcdefg')
'YWJjZGVmZwoK'

Decoding
-----------------------------------------------
It's possible to decode string, encoded with `rebus` to their original values.

Required imports

>>> import rebus

b32decode string

>>> rebus.b32decode('MFRGGZDFMZTQUCQK')
'abcdefg'

b64decode string

>>> rebus.b64decode('YWJjZGVmZwoK')
'abcdefg'

urlsafe_b64encode

>>> rebus.urlsafe_b64decode('YWJjZGVmZwoK')
'abcdefg'

License
===============================================
GPL 2.0/LGPL 2.1

Support
===============================================
For any issues contact me at the e-mail given in the `Author` section.

Author
===============================================
Artur Barseghyan <artur.barseghyan@gmail.com>


CHANGELOG.rst
Release history
=====================================
0.2
-------------------------------------
2013-12-20

- Decoding options added.

0.1
-------------------------------------
2013-12-19

- Initial version.

test.sh
reset
./uninstall.sh
./install.sh
python src/rebus/tests.py

src/rebus/tests.py
from __future__ import print_function

__title__ = 'rebus.tests'
__author__ = 'Artur Barseghyan'
__copyright__ = 'Copyright (c) 2013 Artur Barseghyan'
__license__ = 'GPL 2.0/LGPL 2.1'

import unittest

from six import binary_type

import rebus

PRINT_INFO = True

def print_info(func):
    """
    Prints some useful info.
    """
    if not PRINT_INFO:
        return func

    def inner(self, *args, **kwargs):
        result = func(self, *args, **kwargs)

        print('\n\n%s' % func.__name__)
        print('============================')
        if func.__doc__:
            print('""" %s """' % func.__doc__.strip())
        print('----------------------------')
        if result is not None:
            print(result)
        print('\n++++++++++++++++++++++++++++')

        return result
    return inner

TEXTS = (
    'a',
    'ab',
    'abc',
    'abcd',
    'abcde',
    'abcdef',
    'abcdefg',
    'abcdefgh',
    'abcdefghi',
    'abcdefghij',
    'abcdefghijk',
    'abcdefghijkl',
    'abcdefghijklm',
    'abcdefghijklmn',
    'abcdefghijklmno',
    'abcdefghijklmnop',
    'abcdefghijklmnopq',
    'abcdefghijklmnopqr',
    'abcdefghijklmnopqrs',
    'abcdefghijklmnopqrst',
    'abcdefghijklmnopqrstu',
    'abcdefghijklmnopqrstuv',
    'abcdefghijklmnopqrstuvw',
    'abcdefghijklmnopqrstuvwx',
    'abcdefghijklmnopqrstuvwxy',
    'abcdefghijklmnopqrstuvwxyz',
)

class RebusEncodingTest(unittest.TestCase):
    """
    Encoding tests.
    """
    def setUp(self):
        self.texts = TEXTS
        self.equal_sign = binary_type('='.encode())

    def __encode(self, encoder):
        """
        :param callable encoder:
        :return list:
        """
        flow = []
        for text in self.texts:
            encoded_text = encoder(text)
            self.assertTrue(self.equal_sign not in encoded_text)
            flow.append((text, encoded_text))

        return flow

    @print_info
    def test_01_b32encode(self):
        """
        Test the ``rebus.b32encode``.
        """
        return self.__encode(rebus.b32encode)

    @print_info
    def test_02_b64encode(self):
        """
        Test the ``rebus.b64encode``.
        """
        return self.__encode(rebus.b64encode)

    @print_info
    def test_03_urlsafe_b64encode(self):
        """
        Test the ``rebus.urlsafe_b64encode``.
        """
        return self.__encode(rebus.urlsafe_b64encode)


class RebusDecodingTest(unittest.TestCase):
    """
    Decoding tests.
    """
    def setUp(self):
        self.texts = TEXTS
        self.equal_sign = binary_type('='.encode())

    def __decode(self, encoder, decoder):
        """
        :param callable encoder:
        :param callable decoder:
        :return list:
        """
        flow = []
        for text in self.texts:
            encoded_text = encoder(text)
            self.assertTrue(self.equal_sign not in encoded_text)
            decoded_text = decoder(encoded_text)
            self.assertTrue(decoded_text == text)
            flow.append((text, encoded_text, decoded_text))

        return flow

    @print_info
    def test_01_b32decode(self):
        """
        Test the ``rebus.b32decode``.
        """
        return self.__decode(rebus.b32encode, rebus.b32decode)

    @print_info
    def test_02_b64decode(self):
        """
        Test the ``rebus.b64decode``.
        """
        return self.__decode(rebus.b64encode, rebus.b64decode)

    @print_info
    def test_03_urlsafe_b64encode(self):
        """
        Test the ``rebus.urlsafe_b64decode``.
        """
        return self.__decode(rebus.urlsafe_b64encode, rebus.urlsafe_b64decode)


if __name__ == "__main__":
    # Tests
    unittest.main()


src/rebus/__init__.py
from __future__ import print_function

__title__ = 'rebus'
__version__ = '0.2'
__build__ = 0x000002
__author__ = 'Artur Barseghyan'
__copyright__ = 'Copyright (c) 2013 Artur Barseghyan'
__license__ = 'GPL 2.0/LGPL 2.1'
__all__ = (
    'encode', 'b32encode', 'b64encode', 'urlsafe_b64encode',
    'decode', 'b32decode', 'b64decode', 'urlsafe_b64decode',
    )

import logging
import base64

from six import binary_type, PY3

from rebus.defaults import DEFAULT_SUFFIX

logger = logging.getLogger(__file__)

# ********************************************************
# ********************** Encoding ************************
# ********************************************************

class EncodedText(object):
    """
    Container.
    """
    def __init__(self, text, encoded):
        self.text = text
        self.encoded = encoded

    def __str__(self):
        return str(self.encoded)
    __unicode__ = __str__
    __repr__ = __str__


def encode(encoder, step, text, return_object=False):
    """
    :param callable encoder:
    :param int step:
    :param string text:
    :param bool return_object:
    :return string:
    """
    changed_text = binary_type((text + ((int(len(text) / step) + 1) * step - len(text)) * DEFAULT_SUFFIX).encode())
    encoded_text = encoder(changed_text)
    if return_object:
        return EncodedText(text=changed_text, encoded=encoded_text)
    return encoded_text

def b32encode(text, return_object=False):
    """
    :param string text:
    :param bool return_object:
    :return string:
    """
    return encode(base64.b32encode, 5, text, return_object=return_object)

def b64encode(text, return_object=False):
    """
    :param string text:
    :param bool return_object:
    :return string:
    """
    return encode(base64.b64encode, 3, text, return_object=return_object)

def urlsafe_b64encode(text, return_object=False):
    """
    :param string text:
    :param bool return_object:
    :return string:
    """
    return encode(base64.urlsafe_b64encode, 3, text, return_object=return_object)

# ********************************************************
# ********************** Decoding ************************
# ********************************************************

class DecodedText(object):
    """
    Container.
    """
    def __init__(self, text, decoded):
        self.text = text
        self.decoded = decoded

    def __str__(self):
        return str(self.decoded)
    __unicode__ = __str__
    __repr__ = __str__


def decode(decoder, text, return_object=False):
    """
    :param callable decoder:
    :param string text:
    :param bool return_object:
    :return string:
    """
    decoded_text = decoder(text.decode()).decode().replace(DEFAULT_SUFFIX, '')

    if return_object:
        return DecodedText(text=text, decoded=decoded_text)
    return decoded_text

def b32decode(text, return_object=False):
    """
    :param string text:
    :param bool return_object:
    :return string:
    """
    return decode(base64.b32decode, text, return_object=return_object)

def b64decode(text, return_object=False):
    """
    :param string text:
    :param bool return_object:
    :return string:
    """
    return decode(base64.b64decode, text, return_object=return_object)

def urlsafe_b64decode(text, return_object=False):
    """
    :param string text:
    :param bool return_object:
    :return string:
    """
    if PY3:
        text = text.decode()

    decoded_text = base64.urlsafe_b64decode(text)

    if PY3:
        decoded_text = decoded_text.decode()

    decoded_text = decoded_text.replace(DEFAULT_SUFFIX, '')

    if return_object:
        return DecodedText(text=text, decoded=decoded_text)
    return decoded_text
